README
NAME
    YAML::Syck - Fast, lightweight YAML loader and dumper

SYNOPSIS
        use YAML::Syck;

        # Set this for interoperability with other YAML/Syck bindings:
        # e.g. Load('Yes') becomes 1 and Load('No') becomes ''.
        $YAML::Syck::ImplicitTyping = 1;

        $data = Load($yaml);
        $yaml = Dump($data);

        # $file can be an IO object, or a filename
        $data = LoadFile($file);
        DumpFile($file, $data);

        # A string with multiple YAML streams in it
        $yaml = Dump(@data);
        @data = Load($yaml);

        # Dumping into a pre-existing output buffer
        my $yaml;
        DumpInto(\$yaml, @data);

DESCRIPTION
    This module provides a Perl interface to the libsyck data serialization
    library. It exports the "Dump" and "Load" functions for converting Perl
    data structures to YAML strings, and the other way around.

    NOTE: If you are working with other language's YAML/Syck bindings (such
    as Ruby), please set $YAML::Syck::ImplicitTyping to 1 before calling the
    "Load"/"Dump" functions. The default setting is for preserving
    backward-compatibility with "YAML.pm".

Differences Between YAML::Syck and YAML
  Error handling
    Some calls are designed to die rather than returning YAML. You should
    wrap your calls in eval to assure you do not get unexpected results.

FLAGS
  $YAML::Syck::Headless
    Defaults to false. Setting this to a true value will make "Dump" omit
    the leading "---\n" marker.

  $YAML::Syck::SortKeys
    Defaults to false. Setting this to a true value will make "Dump" sort
    hash keys.

  $YAML::Syck::SingleQuote
    Defaults to false. Setting this to a true value will make "Dump" always
    emit single quotes instead of bare strings.

  $YAML::Syck::ImplicitTyping
    Defaults to false. Setting this to a true value will make "Load"
    recognize various implicit types in YAML, such as unquoted "true",
    "false", as well as integers and floating-point numbers. Otherwise, only
    "~" is recognized to be "undef".

  $YAML::Syck::ImplicitUnicode
    Defaults to false. For Perl 5.8.0 or later, setting this to a true value
    will make "Load" set Unicode flag on for every string that contains
    valid UTF8 sequences, and make "Dump" return a unicode string.

    Regardless of this flag, Unicode strings are dumped verbatim without
    escaping; byte strings with high-bit set will be dumped with backslash
    escaping.

    However, because YAML does not distinguish between these two kinds of
    strings, so this flag will affect loading of both variants of strings.

    If you want to use LoadFile or DumpFile with unicode, you are required
    to open your own file in order to assure it's UTF8 encoded:

      open(my $fh, ">:encoding(UTF-8)", "out.yml");
      DumpFile($fh, $hashref);

  $YAML::Syck::ImplicitBinary
    Defaults to false. For Perl 5.8.0 or later, setting this to a true value
    will make "Dump" generate Base64-encoded "!!binary" data for all
    non-Unicode scalars containing high-bit bytes.

  $YAML::Syck::UseCode / $YAML::Syck::LoadCode / $YAML::Syck::DumpCode
    These flags control whether or not to try and eval/deparse perl source
    code; each of them defaults to false.

    Setting $YAML::Syck::UseCode to a true value is equivalent to setting
    both $YAML::Syck::LoadCode and $YAML::Syck::DumpCode to true.

  $YAML::Syck::LoadBlessed
    Defaults to true. Setting to false will block YAML::Syck from doing ANY
    blessing. This is an interface change since 1.21. The variable name was
    misleading, implying that no blessing would happen when in fact it did.

    Prior to 1.22, setting this to a false value only prevented "Load" from
    blessing tag names that did not begin with "!!perl" or "!perl";.

BUGS
    Dumping Glob/IO values do not work yet.

    Dumping of Tied variables is unsupported.

    Dumping into tied (or other magic variables) with "DumpInto" might not
    work properly in all cases.

CAVEATS
    This module implements the YAML 1.0 spec. To deal with data in YAML 1.1,
    please use the "YAML::XS" module instead.

    The current implementation bundles libsyck source code; if your system
    has a site-wide shared libsyck, it will *not* be used.

    Tag names such as "!!perl/hash:Foo" is blessed into the package "Foo",
    but the "!hs/foo" and "!!hs/Foo" tags are blessed into "hs::Foo". Note
    that this holds true even if the tag contains non-word characters; for
    example, "!haskell.org/Foo" is blessed into "haskell.org::Foo". Please
    use Class::Rebless to cast it into other user-defined packages. You can
    also set the LoadBlessed flag false to disable all blessing.

    This module has a lot of known issues
    <https://rt.cpan.org/Public/Dist/Display.html?Name=YAML-Syck> and has
    only been semi-actively maintained since 2007. If you encounter an issue
    with it probably won't be fixed unless you offer up a patch
    <http://github.com/toddr/YAML-Syck> in Git that's ready for release.

    There are still good reasons to use this module, such as better
    interoperability with other syck wrappers (like Ruby's), or some edge
    case of YAML's syntax that it handles better. It'll probably work
    perfectly for you, but if it doesn't you may want to look at YAML::XS,
    or perhaps at looking another serialization format like JSON.

SEE ALSO
    YAML, JSON::Syck

    <http://www.yaml.org/>

AUTHORS
    Audrey Tang <cpan@audreyt.org>

COPYRIGHT
    Copyright 2005-2009 by Audrey Tang <cpan@audreyt.org>.

    This software is released under the MIT license cited below.

    The libsyck code bundled with this library is released by "why the lucky
    stiff", under a BSD-style license. See the COPYING file for details.

  The "MIT" License
    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
[Changes for 1.23_01 - 2013-02-21]

* Make sure that LoadBlessed=0 blocks all blessing.
* Start listing primary repo as toddr on github ( http://github.com/toddr/YAML-Syck )
* README refreshed via perldoc -t

[Changes for 1.22 - 2012-12-04]

* Tests all green on CPAN Testers. releasing to public.

[Changes for 1.22_01 - 2012-11-27]

* Fix error message typo: s/existant/existent/g
* Fix for non-printable character detection
* Quote if non-printable characters are present
* Change LoadBlessed=0 to not bless under ANY circumstances

[Changes for 1.21 - 2012-09-20]

* Tests all green on CPAN Testers. releasing to public.

[Changes for 1.21_01 - 2012-06-22]

* Synchronize JSON::Syck with YAML::Syck version number
* Add DumpInto functions (YAML+Syck) which dump into
  a provided scalar instead of a newly allocated one
* Modify DumpFile functions to output directly to the
  specified file/filehandle instead of buffering all
  output in memory.
* Avoid modifying numbers into strings when emitting

[Changes for 1.20 (JSON::Syck 0.43) - 2012-02-14]

* Tests all green on CPAN Testers. releasing to public.

[Changes for 1.20_01 (JSON::Syck 0.43_01) - 2012-02-11]

* Add test to assure META.yml is up to date before release
* Fix for RT 70933 - disable check for empty file handle in LoadFile calls
* Remove call to non-existent code YAML::Syck::_is_openhandle.
* Fix for RT 74785 - compiling with -Werror=format-security fails

[Changes for 1.19 (JSON::Syck 0.42 ) - 2011-11-08]

* Do not run t/leak.t on perls prior to 5.8.9 as 5.8.8
  and Devel::Leak apparently do not mix (perl RT #102606).

[Changes for 1.18 (JSON::Syck 0.41 ) - 2011-11-05]

* Promote to stable release.

[Changes for 1.17_02 (JSON::Syck 0.40_02 ) - 2011-11-03]

* perl RT #102606 - Only lvalue subroutines are supposed to return
  the argument to "return" without copying it. Stop relying on a
  pre-5.15 bug. [Father Chrysostomos]

[Changes for 1.17_01 (JSON::Syck 0.40_01 ) - 2011-02-28]

* Break YAML::Syck dependency on Scalar::Util

[Changes for 1.17 (JSON::Syck 0.40 ) - 2010-11-21]

* Change a C99 comment to a C89 comment

* Upgrade ppport.h from 3.13 to 3.19

* Remove inc/Test/Builder/IO/Scalar.pm from MANIFEST

[Changes for 1.16 (JSON::Syck 0.39 ) - 2010-11-21]

* POD formatting fixes contributed by Justin Hunter

[Changes for 1.15 (JSON::Syck 0.38 ) - 2010-09-23]

* perl RT #78008 - The 5.13.5 development release of perl changed how
  regexes are stringified, breaking our tests. Update the tests to
  deal with the new syntax.

  See fb85c04 (http://perl5.git.perl.org/perl.git/commit/fb85c04) in
  perl.git for context.

[Changes for 1.14 (JSON::Syck 0.37 ) - 2010-08-03]
* RT RT60771 - Further refactor for JSON. if a number is a valid
  perl string, quote it. "1" + "1" = "11" in javascript.
* add YAML tests to make sure HEX and octal preserve their information

[Changes for 1.13 (JSON::Syck 0.36 ) - 2010-08-03]
* RT RT60771 - quoted negative integer and quoted zeroes.
  This was breaking JSON since "0" is true in javascript.

[Changes for 1.12 (JSON::Syck 0.35 ) - 2010-08-03]
* The tie test proved to be very unstable for 5.8 and it's not
  relevant for tie actually. I'm converting it to a TODO for 

[Changes for 1.11 (JSON::Syck 0.35 ) - 2010-08-03]
* Tests are now stable. releasing 1.11

[Changes for 1.10_07 (JSON::Syck 0.34 ) - 2010-07-29]
* META.yml missing

[Changes for 1.10_06 (JSON::Syck 0.34 ) - 2010-07-29]
* Load of in-memory file handles not previously tested (was eval'd 
  out of existance) Corrected code and test to behave correctly.
* Perl 5.8 required to YAML-ify code refs
* Perl 5.6 requires {} to print to a scalar GLOB
* use SvNIOK to test for int or float if we know it's not a string
* Attempt again to get 5.8.7 skipped for tie test

[Changes for 1.10_05 (JSON::Syck 0.34 ) - 2010-07-28]
* Skip failing tie test on perl < 5.8.8 && perl >= 5.7.0x

[Changes for 1.10_04 (JSON::Syck 0.34 ) - 2010-07-19]
* RT 23850 - META.yml of DMAKI/DateTime-Format-Japanese-0.01.tar.gz
  cannot be parsed
* RT 34073 - TODO Tests - Parsing YAML without separator
* RT 20969 - Document die behavior differences with YAML

[Changes for 1.10_03 (JSON::Syck 0.34 ) - 2010-07-19]
* TODO tests to cover noted deficiencies in RT #'s
  26182, 54780, 49404, 42390
* RT 23909 - Loaded self-refererring document is corrupted

[Changes for 1.10_02 (JSON::Syck 0.34 ) - 2010-07-19]
* RT 50227 - Logic bug in deciding what numbers to quote led to some
  interesting corner cases. Now only 9digit integers are supported
  as unquoted. Everything else will be forced to quote
* RT 59458 - convert \%d to \%ld in code to prevent compiler warnings

[Changes for 1.10_01 (JSON::Syck 0.34) - 2010-07-15]
* LoadFile aborts loading an empty file
* RT 50227 - Not quoting strings when data has been sorted numerically
  before being passed to JSON::Syck::Dump
* RT 48327 - use 3 arg form of open() instead of the 2 argument form
* Add TODO tests for RT 52432 - '... X' breaks round trip
* RT 59432 - use lexical file handles for read/write to files
* RT 37236 - Re-enable JSON-basic tests for JSON >= 2.11
* RT 47944 - back quote is a reserved character
* RT 36288 - ability to disable blessing datastructures on load
* Tone down the WARNING section in YAML::Syck's documentation that was
  added in 1.08

[Changes for 1.10 (JSON::Syck 0.33) - 2010-06-06]

1.09 was released with version 0.97 of Module::Install. It had a
critical bug that I reported which broke installations on perl 5.8.3
and older.

This release is equivalent to 1.09 aside from bumping Module::Install
to 0.99 which fixed the bug.

[Changes for 1.09 (JSON::Syck 0.32) - 2010-05-29]

* Exactly equivalent to 1.08_01, aside from bumping the version
  numbers

[Changes for 1.08_01 (JSON::Syck 0.31_01) - 2010-05-23]

* Reset the hash iterator after dumping hashes. Solves RT #54167 and
  the duplicate RT #34166.

* RT #39572: Only print 0.60 compatibility warning if 0.60 or earlier
  is installed.


lib/YAML/Syck.pm
package YAML::Syck;
# See documentation after the __END__ mark.

use strict;
use vars qw(
    @ISA @EXPORT @EXPORT_OK $VERSION
    $Headless $SortKeys $SingleQuote
    $ImplicitBinary $ImplicitTyping $ImplicitUnicode 
    $UseCode $LoadCode $DumpCode
    $DeparseObject $LoadBlessed
);
use 5.006;
use Exporter;

BEGIN {
    $VERSION   = '1.23_01';
    @EXPORT    = qw( Dump Load DumpFile LoadFile );
    @EXPORT_OK = qw( DumpInto );
    @ISA       = qw( Exporter );

    $SortKeys = 1;
    $LoadBlessed = 1;

    local $@;
    eval {
        require XSLoader;
        XSLoader::load(__PACKAGE__, $VERSION);
        1;
    } or do {
        require DynaLoader;
        push @ISA, 'DynaLoader';
        __PACKAGE__->bootstrap($VERSION);
    };

}

use constant QR_MAP => {
    ''   => sub { qr{$_[0]}     }, 
    x    => sub { qr{$_[0]}x    }, 
    i    => sub { qr{$_[0]}i    }, 
    s    => sub { qr{$_[0]}s    }, 
    m    => sub { qr{$_[0]}m    }, 
    ix   => sub { qr{$_[0]}ix   }, 
    sx   => sub { qr{$_[0]}sx   }, 
    mx   => sub { qr{$_[0]}mx   }, 
    si   => sub { qr{$_[0]}si   }, 
    mi   => sub { qr{$_[0]}mi   }, 
    ms   => sub { qr{$_[0]}sm   }, 
    six  => sub { qr{$_[0]}six  }, 
    mix  => sub { qr{$_[0]}mix  }, 
    msx  => sub { qr{$_[0]}msx  }, 
    msi  => sub { qr{$_[0]}msi  }, 
    msix => sub { qr{$_[0]}msix }, 
};

sub __qr_helper {
    if ($_[0] =~ /\A  \(\?  ([ixsm]*)  (?:-  (?:[ixsm]*))?  : (.*) \)  \z/x) {
        my $sub = QR_MAP()->{$1} || QR_MAP()->{''};
        &$sub($2);
    }
    else {
        qr/$_[0]/;
    }
}

sub Dump {
    $#_ ? join('', map { YAML::Syck::DumpYAML($_) } @_)
        : YAML::Syck::DumpYAML($_[0]);
}

sub Load {
    if (wantarray) {
        my ($rv) = YAML::Syck::LoadYAML($_[0]);
        @{$rv};
    }
    else {
        @_ = $_[0];
        goto &YAML::Syck::LoadYAML;
    }
}

sub _is_glob {
    my $h = shift;
    
    return 1 if(ref($h) eq 'GLOB');
    return 1 if(ref(\$h) eq 'GLOB');
    return 1 if(ref($h) =~ m/^IO::/);

    return;    
}

sub DumpFile {
    my $file = shift;
    if ( _is_glob($file) ) {
        for (@_) {
            my $err = YAML::Syck::DumpYAMLFile($_, $file);
            if ($err) {
                $! = 0+$err;
                die "Error writing to filehandle $file: $!\n";
            }
        }
    } else {
        open(my $fh, '>', $file) or die "Cannot write to $file: $!";
        for (@_) {
            my $err = YAML::Syck::DumpYAMLFile($_, $fh);
            if ($err) {
                $! = 0+$err;
                die "Error writing to file $file: $!\n";
            }
        }
        close $fh
            or die "Error writing to file $file: $!\n";
    }
    return 1;
}

sub LoadFile {
    my $file = shift;
    if ( _is_glob($file) ) {
      Load(do { local $/; <$file> });
    }
    else {
      if(!-e $file || -z $file) {
	die("'$file' is empty or non-existent");
      }
        open(my $fh, '<', $file) or die "Cannot read from $file: $!";
        Load(do { local $/; <$fh> });
    }
}

sub DumpInto {
    my $bufref = shift;
    (ref $bufref) or die "DumpInto not given reference to output buffer\n";
    YAML::Syck::DumpYAMLInto($_, $bufref) for @_;
    1;
}

1;

__END__
=pod

=head1 NAME 

YAML::Syck - Fast, lightweight YAML loader and dumper

=head1 SYNOPSIS

    use YAML::Syck;

    # Set this for interoperability with other YAML/Syck bindings:
    # e.g. Load('Yes') becomes 1 and Load('No') becomes ''.
    $YAML::Syck::ImplicitTyping = 1;

    $data = Load($yaml);
    $yaml = Dump($data);

    # $file can be an IO object, or a filename
    $data = LoadFile($file);
    DumpFile($file, $data);

    # A string with multiple YAML streams in it
    $yaml = Dump(@data);
    @data = Load($yaml);

    # Dumping into a pre-existing output buffer
    my $yaml;
    DumpInto(\$yaml, @data);

=head1 DESCRIPTION

This module provides a Perl interface to the B<libsyck> data serialization
library.  It exports the C<Dump> and C<Load> functions for converting
Perl data structures to YAML strings, and the other way around.

B<NOTE>: If you are working with other language's YAML/Syck bindings
(such as Ruby), please set C<$YAML::Syck::ImplicitTyping> to C<1> before
calling the C<Load>/C<Dump> functions.  The default setting is for
preserving backward-compatibility with C<YAML.pm>.

=head1 Differences Between YAML::Syck and YAML

=head2 Error handling

Some calls are designed to die rather than returning YAML. You should wrap
your calls in eval to assure you do not get unexpected results.

=head1 FLAGS

=head2 $YAML::Syck::Headless

Defaults to false.  Setting this to a true value will make C<Dump> omit the
leading C<---\n> marker.

=head2 $YAML::Syck::SortKeys

Defaults to false.  Setting this to a true value will make C<Dump> sort
hash keys.

=head2 $YAML::Syck::SingleQuote

Defaults to false.  Setting this to a true value will make C<Dump> always emit
single quotes instead of bare strings.

=head2 $YAML::Syck::ImplicitTyping

Defaults to false.  Setting this to a true value will make C<Load> recognize
various implicit types in YAML, such as unquoted C<true>, C<false>, as well as
integers and floating-point numbers.  Otherwise, only C<~> is recognized to
be C<undef>.

=head2 $YAML::Syck::ImplicitUnicode

Defaults to false.  For Perl 5.8.0 or later, setting this to a true value will
make C<Load> set Unicode flag on for every string that contains valid UTF8
sequences, and make C<Dump> return a unicode string.

Regardless of this flag, Unicode strings are dumped verbatim without escaping;
byte strings with high-bit set will be dumped with backslash escaping.

However, because YAML does not distinguish between these two kinds of strings,
so this flag will affect loading of both variants of strings.

If you want to use LoadFile or DumpFile with unicode, you are required to open
your own file in order to assure it's UTF8 encoded:

  open(my $fh, ">:encoding(UTF-8)", "out.yml");
  DumpFile($fh, $hashref);

=head2 $YAML::Syck::ImplicitBinary

Defaults to false.  For Perl 5.8.0 or later, setting this to a true value will
make C<Dump> generate Base64-encoded C<!!binary> data for all non-Unicode
scalars containing high-bit bytes.

=head2 $YAML::Syck::UseCode / $YAML::Syck::LoadCode / $YAML::Syck::DumpCode

These flags control whether or not to try and eval/deparse perl source code;
each of them defaults to false.

Setting C<$YAML::Syck::UseCode> to a true value is equivalent to setting
both C<$YAML::Syck::LoadCode> and C<$YAML::Syck::DumpCode> to true.

=head2 $YAML::Syck::LoadBlessed

Defaults to true. Setting to false will block YAML::Syck from doing ANY
blessing. This is an interface change since 1.21. The variable name was
misleading, implying that no blessing would happen when in fact it did.

Prior to 1.22, setting this to a false value only prevented C<Load> from
blessing tag names that did not begin with C<!!perl> or C<!perl>;.

=head1 BUGS

Dumping Glob/IO values do not work yet.

Dumping of Tied variables is unsupported.

Dumping into tied (or other magic variables) with C<DumpInto> might not work
properly in all cases.

=head1 CAVEATS

This module implements the YAML 1.0 spec.  To deal with data in YAML 1.1, 
please use the C<YAML::XS> module instead.

The current implementation bundles libsyck source code; if your system has a
site-wide shared libsyck, it will I<not> be used.

Tag names such as C<!!perl/hash:Foo> is blessed into the package C<Foo>, but
the C<!hs/foo> and C<!!hs/Foo> tags are blessed into C<hs::Foo>.  Note that
this holds true even if the tag contains non-word characters; for example,
C<!haskell.org/Foo> is blessed into C<haskell.org::Foo>.  Please use
L<Class::Rebless> to cast it into other user-defined packages. You can also
set the LoadBlessed flag false to disable all blessing.

This module has L<a lot of known
issues|https://rt.cpan.org/Public/Dist/Display.html?Name=YAML-Syck>
and has only been semi-actively maintained since 2007. If you
encounter an issue with it probably won't be fixed unless you L<offer
up a patch|http://github.com/toddr/YAML-Syck> in Git that's ready for
release.

There are still good reasons to use this module, such as better
interoperability with other syck wrappers (like Ruby's), or some edge
case of YAML's syntax that it handles better. It'll probably work
perfectly for you, but if it doesn't you may want to look at
L<YAML::XS>, or perhaps at looking another serialization format like
L<JSON>.

=head1 SEE ALSO

L<YAML>, L<JSON::Syck>

L<http://www.yaml.org/>

=head1 AUTHORS

Audrey Tang E<lt>cpan@audreyt.orgE<gt>

=head1 COPYRIGHT

Copyright 2005-2009 by Audrey Tang E<lt>cpan@audreyt.orgE<gt>.

This software is released under the MIT license cited below.

The F<libsyck> code bundled with this library is released by
"why the lucky stiff", under a BSD-style license.  See the F<COPYING>
file for details.

=head2 The "MIT" License

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.

=cut
